crates/sui-core/src/authority/authority_store_pruner.rs

```diff
@@ -289,8 +289,12 @@ impl AuthorityStorePruner {
             .get_archive_watermark()
             .await?
             .unwrap_or(u64::MAX);
-        let highest_pruned_checkpoint = perpetual_db.get_highest_pruned_checkpoint()?;
-        info!("Latest archived checkpoint: {latest_archived_checkpoint}, highest pruned checkpoint: {highest_pruned_checkpoint}");
+        let objects_pruned_checkpoint = if config.num_epochs_to_retain != u64::MAX {
```

Comment:
so in this way only configurations with explicit `num_epochs_to_retain == u64::MAX` (which is not default so has to be enabled explicitly) will be able to prune transactions without objects.
For all other setups, the same guarantee remains. Transaction pruner won't process checkpoint x before the object pruner